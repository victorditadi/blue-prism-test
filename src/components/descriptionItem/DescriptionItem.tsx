import { sliceAndUpperCaseString } from 'helpers';
import React from 'react';

import {
  Container,
  DescriptionType,
  DescriptionValueContainer,
  TooltipContainer,
  DescriptionValue,
} from './styles';

const DescriptionItem = ({
  itemKey,
  itemValue,
  type,
}: {
  itemKey: string;
  itemValue?: string | number | boolean | undefined;
  type: string;
}) => {
  const valueIsString = typeof itemValue === 'string';
  const tooltipOn = valueIsString ? itemValue.length > 10 : false;
  const shortText =
    valueIsString && tooltipOn ? `${itemValue.slice(0, 10)}...` : itemValue?.toString();

  return (
    <Container data-testid={`${type}-${itemKey}-container`}>
      <DescriptionType data-testid={`${type}-${itemKey}-type`}>
        {sliceAndUpperCaseString(itemKey)}:
      </DescriptionType>
      <DescriptionValueContainer data-testid={`${type}-${itemKey}-value`}>
        <DescriptionValue data-testid="short-text">{shortText}</DescriptionValue>
        {tooltipOn && <TooltipContainer data-testid="tooltip">{itemValue}</TooltipContainer>}
      </DescriptionValueContainer>
    </Container>
  );
};

export { DescriptionItem };
