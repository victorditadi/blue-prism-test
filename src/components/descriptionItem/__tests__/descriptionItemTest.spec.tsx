import { fireEvent, render } from '@testing-library/react';

import Theme from 'css/theme';
import { DescriptionItem } from '../DescriptionItem';

describe('Component: DescriptionItem', () => {
  const renderComponent = ({ text }: { text: string }) =>
    render(
      <Theme>
        <DescriptionItem
          key={`test-testType-container`}
          data-testid={`test-testType-container`}
          type={'test'}
          itemValue={text}
          itemKey={'testWithCamelCase'}
        />
      </Theme>
    );

  it('should render DescriptionItem component', () => {
    const { asFragment } = renderComponent({ text: 'small text' });
    expect(asFragment()).toMatchSnapshot();
  });

  it('should contain Text', () => {
    const { queryByTestId } = renderComponent({ text: 'small text' });
    expect(queryByTestId('short-text')).toBeTruthy();
  });

  it('should slice Text if Lenght > 10', () => {
    const { queryByTestId } = renderComponent({
      text: 'Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua.',
    });

    expect(queryByTestId('short-text')?.textContent).toStrictEqual('Lorem ipsu...');
  });

  it('should not trigger tooltip outside mouseover', async () => {
    const { queryByTestId } = renderComponent({
      text: 'Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua.',
    });

    const tooltipContainer = queryByTestId('tooltip');
    expect(tooltipContainer).not.toBeVisible();
  });

  it('should trigger tooltip when mouseover in a long text', async () => {
    const { queryByTestId } = renderComponent({
      text: 'Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua.',
    });
    const shortText = queryByTestId('short-text');
    shortText && fireEvent.mouseOver(shortText);

    const tooltipContainer = queryByTestId('tooltip');
    expect(tooltipContainer).toBeVisible;
  });

  it('should slice and uppercase type description', () => {
    const { queryByTestId } = renderComponent({ text: '' });
    const descriptionContainer = queryByTestId(`test-testWithCamelCase-type`);
    expect(descriptionContainer?.textContent).toStrictEqual('Test With Camel Case:');
  });
});
