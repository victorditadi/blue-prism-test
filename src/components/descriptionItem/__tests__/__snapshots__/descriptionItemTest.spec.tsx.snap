// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component: DescriptionItem should render DescriptionItem component 1`] = `
<DocumentFragment>
  .c0 {
  -webkit-flex: 50%;
  -ms-flex: 50%;
  flex: 50%;
  font-family: Robotosans-serif;
  font-size: 1em;
  margin-top: 8px;
}

.c1 {
  font-family: Robotosans-serif;
  font-weight: bold;
  font-size: 1em;
}

.c2 {
  font-size: 1em;
  font-family: Robotosans-serif;
}

<span
    class="c0"
    data-testid="test-testWithCamelCase-container"
  >
    <span
      class="c1"
      data-testid="test-testWithCamelCase-type"
    >
      Test With Camel Case:
    </span>
    <div
      class=""
      data-testid="test-testWithCamelCase-value"
    >
      <span
        class="c2"
        data-testid="short-text"
      >
        small text
      </span>
    </div>
  </span>
</DocumentFragment>
`;
